### DISCUSSION TITLE
Prevent upwards Y force applying after force to X and Z on slope

### DISCUSSION THREAD

**Discussion by quaternion1337** [enable3d/enable3d#154] [Help / Questions] [answered]
Hi guys

First the problem. 

Whenever i use the character to run over the little hill seen in this picture.
![image](https://user-images.githubusercontent.com/12425208/113588218-23e8fb80-9630-11eb-94ea-83de7bb9b656.png)

He flies upwards.
![image](https://user-images.githubusercontent.com/12425208/113588350-4f6be600-9630-11eb-8659-f3bf89c3802f.png)

I've just applied the classical movement behavior like this.
```
   private _Get_Speed_Velocity(timeElapsedSeconds){
        let speed = timeElapsedSeconds * 1000;
        if (speed > 10) speed = 10;
        
        const camera_direction = new THREE.Vector3();
        this.CharacterModel._camera.getWorldDirection(camera_direction);
        const angle_radians = Math.atan2(camera_direction.x, camera_direction.z);

        let x =  Math.sin(angle_radians) * speed;
        let y = this.CharacterModel._model.body.velocity.y; 
        let z = Math.cos(angle_radians) * speed;

        switch(true){
            case Data.inputs.forward:
                x = x;
                y = y;
                z = z;
                break;
            case Data.inputs.backward:
                x = -x;
                y = y;
                z = -z;
                break;
            default:
                x = 0;
                y = y;
                z = 0;
                break;
        }

        return {x, y: y, z};
    }
```

And just used it like this
```
        const speed_velocity = this._Get_Speed_Velocity(timeElapsedSeconds);
        this.CharacterModel._model.body.setVelocity(speed_velocity.x, speed_velocity.y, speed_velocity.z);
```

I've tried changing mass, gravity, friction and shapes. (Mass doesn't even seem to have any effect?)

```
            this.options = {
                addChildren: false,
                shape: 'capsule',
                radius: 40,
                height: 115,
                offset: {y: -0.8}
            }
            Data.physics.add.existing(this.CharacterModel._model, this.options);
            this.CharacterModel._model.body.setCollisionFlags(0);
            this.CharacterModel._model.body.setAngularFactor(0, 0, 0)
            this.CharacterModel._model.body.setFriction(0.8);
            this.CharacterModel._model.body.setCcdMotionThreshold(1e-7);
            this.CharacterModel._model.body.setCcdSweptSphereRadius(1);
```

How do i prevent it acting like a ball? 
I just dont want that massive upwards force to Y whenever a force to X and Z occur on a slope.

**Comment by quaternion1337** [comment]
Same issue actually happens in the demo on the official webpage.

Run to this stone over here:
![image](https://user-images.githubusercontent.com/81978012/113615876-b9968200-9654-11eb-9902-1dbd15c438d9.png)

Try and jump when you run upwards the stone and you will find yourself flying way higher than you do with the normal jump.
![image](https://user-images.githubusercontent.com/81978012/113616082-f4001f00-9654-11eb-8c53-e7c5ed8f7a9d.png)

**Comment by yandeu** [comment]
I don't see where you are applying force (`body.applyForce()`).

If you want the boxMan to jump less high in the Medieval Example, try:

```ts
const y = this.man.body.velocity.y
const force = 6
this.man.body.applyForceY(y > 0.01 ? 6 - y : 6)
```

**Reply by quaternion1337** [reply]
Thanks for the answer.

It's not the fact that he should jump less high. 
It's the fact that bodies are acting like cars and objects with a slope at 45 degress function like a ramp for the object.

I'm just setting the velocity with setVelocity. I've tried with applyForceX and applyForceY but the same thing happen. 

I'm no artist but i've tried to draw the problem
![image](https://user-images.githubusercontent.com/81978012/113622467-565d1d80-965d-11eb-83dd-ca390262d91d.png)
I'm after the behavior of the bottom illustration

The only way i found out to obtain this is to set gravity to (0, -9.82*4, 0) which is not ideal

**Comment by quaternion1337** [comment, marked as answer]
I found a simple solution

```
  const str_vector = new THREE.Vector3(x,0,z);
  const mov_vector = new THREE.Vector3(x,y,z);
  const angle = (mov_vector.angleTo(str_vector) * 180) / Math.PI;
  if(angle > 30){
      y = y/3;
  }
```

This will slow down the force being applied when the object is moving upwards
